server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -77,65 +87,212 @@
+            for (DiscoveryNode node : allNodes) {
+                if (randomBoolean()) {
+                    final long disconnectionTime = randomLongBetween(0, 120000);
+                    maxDisconnectionTime = Math.max(maxDisconnectionTime, disconnectionTime);
+                    deterministicTaskQueue.scheduleAt(disconnectionTime, new Runnable() {
+                        @Override
+                        public void run() {
+                            transportService.disconnectFromNode(node);
+                        }
+
+                        @Override
+                        public String toString() {
+                            return "scheduled disconnection of " + node;
+                        }
+                    });
+                }
             }
-            service.new ConnectionChecker().run();
         }
 
+        runTasksUntil(deterministicTaskQueue, maxDisconnectionTime);
+
         // disable exceptions so things can be restored
         transport.randomConnectionExceptions = false;
-        service.new ConnectionChecker().run();
-        assertConnectedExactlyToNodes(event.state());
+        logger.info("renewing connections");
+        runTasksUntil(deterministicTaskQueue, maxDisconnectionTime + reconnectIntervalMillis);
+        assertConnectedExactlyToNodes(targetNodes);
+    }
+
+    public void testOnlyBlocksOnConnectionsToNewNodes() throws Exception {
+        final NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
+
+        // connect to one node
+        final DiscoveryNode node0 = new DiscoveryNode("node0", buildNewFakeTransportAddress(), Version.CURRENT);
+        final DiscoveryNodes nodes0 = DiscoveryNodes.builder().add(node0).build();
+        final PlainActionFuture<Void> future0 = new PlainActionFuture<>();
+        service.connectToNodes(nodes0, () -> future0.onResponse(null));
+        future0.actionGet();
+        assertConnectedExactlyToNodes(nodes0);
+
+        // connection attempts to node0 block indefinitely
+        final CyclicBarrier connectionBarrier = new CyclicBarrier(2);
+        try {
+            nodeConnectionBlocks.put(node0, connectionBarrier::await);
+            transportService.disconnectFromNode(node0);
+
+            // can still connect to another node without blocking
+            final DiscoveryNode node1 = new DiscoveryNode("node1", buildNewFakeTransportAddress(), Version.CURRENT);
+            final DiscoveryNodes nodes1 = DiscoveryNodes.builder().add(node1).build();
+            final DiscoveryNodes nodes01 = DiscoveryNodes.builder(nodes0).add(node1).build();
+            final PlainActionFuture<Void> future1 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future1.onResponse(null));
+            future1.actionGet();
+            assertConnectedExactlyToNodes(nodes1);
+
+            // can also disconnect from node0 without blocking
```

Comment:
I think we would need to wait for the barrier on disconnects to validate this?

Reply:
The connection attempt is still waiting on the barrier; the point is that `disconnectFromNodesExcept()` returns immediately rather than waiting for the connection attempt to complete.